### PR TITLE
[DPE-4656] add TLS CA rotation routine

### PR DESCRIPTION
**Pull request by reneradoi**
## Issue
When a new TLS certificate authority (CA) certificate is issued, the opensearch-operator should add this new CA to all its units and request new certificates. The new certificates (including the CA certificate) should be distributed to all OpenSearch nodes in a rolling restart manner, without downtime to the entire cluster.

Due to limitations on the self-signed-certificates operator it is not possible to:
- get a notice if a CA certificate is about to expire
- request a new CA when the current one is about to or has expired
- request an intermediate CA and sign future certificates with it

There is currently no support for renewing a root / CA certificate on the self-signed-certificates operator.  A new root / CA certificate will only be generated and issued if the common_name of the CA changes.

We have decided to implement the logic in that way that we check each certificate if it includes a new CA. If so, we store the new CA and initiate the CA rotation workflow on OpenSearch.

## Solution

This PR implements the following workflow:
- check each `CertificateAvailableEvent` if it includes a new CA
- add the new CA to the truststore
- add a notice `tls_ca_renewing` to the unit's peer data
- initiate a restart of OpenSearch (using the locking mechanism to coordinate cluster availability during the restart)
- after restarting, add a notice `tls_ca_renewed` to the unit's peer data
- when the restart is done on all of the cluster nodes, request new TLS certificates and apply them to the node

During the phase of renewing the CA, all incoming `CertificateAvailableEvents` will be deferred in order to avoid incompatibilites in communication between the nodes.

Please also see the flow of events and actions that has been documented here: https://github.com/canonical/opensearch-operator/wiki/TLS-CA-rotation-flow

## Notes
- There is a dependency to https://github.com/canonical/opensearch-operator/pull/367 because during the rolling restart when the CA is rotated it is very likely that the voting exclusion issue shows up (at least in 3-node-clusters). Therefore the integration test is currently running only with two nodes. Once the voting exclusions issue is resolved, this can be updated to the usual three nodes.
- Due to an upstream bug with JDK it is necessary to use TLS v1.2 (more details see https://github.com/opensearch-project/security/issues/3299).
- This PR introduces a method to append configuration to the jvm options file of OpenSearch (used to set TLS config to v1.2).

### CODE REVIEW

File: lib/charms/opensearch/v0/opensearch_secrets.py
Comment (reneradoi) [inline lib/charms/opensearch/v0/opensearch_secrets.py:122]: Updating the TLS certs after a `secret_change` event has been removed. New TLS certs should only be processed by the `certificate_available` event in order to make sure new CA's are recognized and the workflow for rotating the CA and renewing the certificates is being handled correctly.

File: lib/charms/opensearch/v0/opensearch_secrets.py
Comment (reneradoi) [inline lib/charms/opensearch/v0/opensearch_secrets.py:116]: Updating the TLS certs after a secret_change event has been removed. New TLS certs should only be processed by the certificate_available event in order to make sure new CA's are recognized and the workflow for rotating the CA and renewing the certificates is being handled correctly.

File: lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py:586]: is this needed outside of the CA / Admin cert rotation workflow in the main orchestrator? 
I was under the impression that the TLS workflow is independent from this and would have started without being explicitly requested in case of a new large deployment.

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_tls.py:215]: Should we defer here? or how do we react if the storing of the new CA didn't succeed?

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_tls.py:330]: Could you explain this?

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_tls.py:522]: Would we not save a potential deferral in the current unit if we always saved the admin certificate if available?

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_tls.py:853]: nit:
```suggestion
        
        return False
```

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks René, great work!

I ran the code, the CA rotation completes successfully. However:

- I'm unable to complete the certificate rotation workflow based on the new CA with the following exceptions:
```
unit-main-0: 17:58:43 DEBUG unit.main/0.juju-log opensearch-peers:0: Getting secret app:admin-password
unit-main-0: 17:58:43 DEBUG unit.main/0.juju-log opensearch-peers:0: Starting new HTTPS connection (1): 10.122.32.151:9200
unit-main-0: 17:58:43 DEBUG unit.main/0.juju-log opensearch-peers:0: Error when checking if host 10.122.32.151 is up: HTTP error self.response_code=None
self.response_text="HTTPSConnectionPool(host='10.122.32.151', port=9200): Max retries exceeded with url: / (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1007)')))"
```

- The certificates_relation_changed hook errors once immediately after the CA rotation starts

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:387]: Same as below:

This caused trouble with the locking mechanism: If the operator tries to connect with Opensearch using the "new" admin certificate with the already "new" CA, it can't connect and then at some point the leader just used the peer databag lock to restart upon CA-renewal, independent if another unit may already be restarting. I could fix this by not storing the admin-certs during an ongoing CA-renewal, but this can only be controlled by deferring the `CertificateAvailableEvent`. Storing admin-certs on other occasions (like `peer-relation changed/created) was destroying this deferral-approach, that's why I removed it from these handlers and decided to only store when really sure it's fine.

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:548]: Good find, I guess this could be removed (oversaw it on cleaning up).

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:743]: That's correct. It's guaranteed by checking `if self.tls.is_fully_configured()`, which checks (via `all_tls_resources_stored()`) that all three certs are there and they all have the current (aka "new") CA.

[Here's](https://github.com/canonical/opensearch-operator/blob/DPE-4656-implement-CA-replacement-routine/lib/charms/opensearch/v0/opensearch_tls.py#L662) the logic for that.

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1044]: The flow is:

get new CA -> set `tls_ca_renewing` -> restart -> `post_start_init` -> set `tls_ca_renewed` -> request new certs -> get new certs -> `on_tls_conf_set` -> delete both `tls_ca_renewing` and `tls_ca_renewed`.

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1047]: Yes, will do so.

File: lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py:586]: This fixed an integration test we have for plugin-management in large deployments. As I did not test it for large deployments myself yet (this will be the scope of the next task for this pulse), I didn't see if working outside of the test.

It fixed the following case: Followers get the new CA from the main orchestrator via `relation-changed`, but didn't receive their new certs yet. I could not yet investigate why they didn't receive/update the certs, but requesting new ones helped here.

This could be subject of change when I work on the large-deployment-task.

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_tls.py:215]: Yes, I will add that.

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_tls.py:330]: I found while investigating another task, will copy the answer from [the PR](https://github.com/canonical/opensearch-operator/pull/408):

```
[...] a bugfix for renewing the expiring certificates, or better creating the certificate signing requests: When initially creating them, they do not include unique-id in the subject. But on renewal this is included, which causes the subject of the renewed certificate to be different from the initial one.
```

Even though the `issuer` will change on CA-rotation, subjects should stay the same if the unit itself didn't change.

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_tls.py:522]: No, this causes trouble in the communication with Opensearch. If we already have the new pem-file but the new `unit-http` cert was not loaded yet, verification of http-requests against any API will fail.

Therefore we update the pem-file only once the certs have been updated (in the `on_tls_conf_set` method).

**Comment by reneradoi** [comment]
> * I'm unable to complete the certificate rotation workflow based on the new CA with the following exceptions:

I also saw this when adding continuous writes to the integration test. The problem there was addressed in [this commit](https://github.com/canonical/opensearch-operator/pull/353/commits/bb98bc2a8bad3dc0740160f9b6e43be82098bf59). Now the certificate renewal after CA-rotation completes, as can be seen in the [integration test](https://github.com/canonical/opensearch-operator/actions/runs/10512550958/job/29130360362).

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: Done.

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1046]: Done.

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_tls.py:216]: Done.

**Review by juditnovak — COMMENTED** [review]
The one main change I'd like to request is to please **add testsing**.

We're adding here a _particularly_ complex change as a result of good 1.5 months' work with more than 120 commits. 

Even if we want to provide integration tests separateley (I assume 1 test, happy path, checking main functionality), we have a good number of isolated "buliding bricks" (methods that should be unit tested). Typicaly the issuer mismatch and so on.
I assume Integration Tests won't get a chance to test a high amount of this new code (typically when anything goes wrong, secret can't be saved, wrong cert was received, etc.)

Main parts of the workflow also should have unit tests (like the charm responds correctly when a new CA cert if provideed, etc.)

I'm 100% sure that the workflow is correct right now. (It's REALLY impressive.)

However given the particularly high complexity (spread across a number of modules) make it particularly difficult to **maintain**. It would be important to have a security layer around the "core moves" of this machinery, that would trigger a test failure if not working correctly. There's a good number of scenarios here, where the input or "start state" conditions can be perfectly described together with the expected state transition from charm's side. ("If we have this, and received that in the event, we want the charm to go this way, OR add this/that to the databag, etc.).

The core workflow transitions (described [here](https://github.com/canonical/opensearch-operator/pull/353/files#r1728650417)) surely should be simulated in unittests. 

I would also add a unittest around the main deliverable (we always have a valid CA certificate available!).

These tests don't only provide security, but also describe the expected behavior (i.e. provide a way of description, "documentation") of the detailed, precise behavior we expect from the charm. 

Me, as a Developer would have a hard time to sleep at night if I had to alter logic in this workflow -- with the current test coverage ;-)  (Or one single Integration Test walking  through the happy path.)

**Review by juditnovak — CHANGES_REQUESTED** [review]
Sry, I was meant to request changes instead of commenting

**Comment by juditnovak** [comment]
To clarify my point (after an offline chat): I'm struggling to approve these changes with **a single happy path** testing.

In case we prefer to move towards Integration Tests instead of Unit Tests, I'm very happy to approve for Integration Tests that verify corner cases, "unhappy" paths etc.
My principal is against committing significant amount of code with NO more test coverage than a 'happy path'. 
If we can demonstrate a 80-90% test coverage for the workflow above in Integration Tests, I'm equally happy. 

(I suggest Unit Tests to verify _specific_ parts of the workflow as they 
 - are lightweight 
 - make it simple to simulate corner cases
 - measurable (i.e. we know if we miss something)

I'm not at all against elaborate Ingetgration Testing with failure cases, but we should keep in mind that
 - pipelines will get heavier
 - return is slower
 - development time is significantly longer
 )
 
 I agree that Integrtaion Testing is a WAY more important than Unit Testng. We are NOT arguing on that.
 (If we were missing Integration Testing for the feature, I'd be equally unhappy ;-) )
 
 _My point is that we need more than a happy path testing. (I don't mind with what method :-) ) We need test coverage for new code._
 
 Equally, it's important to emhapsize: I'm _strictly_ asking for Unit Tests for NEW code/workflows. 
 (NOT to fill in old gaps in Opensearch testing. Please no mis-understandings on that.)

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:744]: Comment before `remove_old_ca()` was added.

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1044]: Comment lines have been added to the `CertificateAvailable` handler in `opensearch_tls`, where the whole workflow is initiated. Other than that there's also [this documentation](https://github.com/canonical/opensearch-operator/wiki/TLS-CA-rotation-flow).

File: lib/charms/opensearch/v0/opensearch_config.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_config.py:115]: There's an unfixed bug in OpenJDK for TLSv1.3 that hits when restarting Opensearch with the updated truststore (where the new CA was added): 
- https://github.com/opensearch-project/security/issues/3299
- https://bugs.openjdk.org/browse/JDK-8221218

The workaround to mitigate this is to use TLSv1.2 instead.

File: lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py:869]: In large deployments it can happen that a non-main-orchestrator is up before the main-orchestrator provides the admin-tls data. If this is the case, the non-main-orchestrator should be blocked until the main-orchestrator provides the data. If this doesn't happen, it could be that the non-main-orchestrator creates an own truststore password that will later be overwritten.

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_tls.py:225]: Thank you for that hint, this `defer()` was indeed not necessary. I've removed it and re-tested the workflow.

Regarding the "lot of defer()-ing": I've tried my best to remove from everywhere where not needed. Still, quite a few deferrals remain. This is due to dependencies, e.g. cert files must not be updated until all nodes have the new CA.

**Comment by juditnovak** [comment]
Our scheduled tests have an 56% test coverage
image

...while this PR has only an 55% test coverage
image

**Comment by juditnovak** [comment]
I have a constructive idea :-)

How about me adding 1-2 unittests to your PR? :-) 
Could this be a good solution for you? 
I checked the TLS unittest module, I beieve I could do it in a short time. 
Thus no contenxt switch on your side (as I know you've moved on to a next heavy topic).
I'd have an hour or so Friday morning to do this. Deal? :-)

File: lib/charms/opensearch/v0/opensearch_base_charm.py
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:398]: Okay let's please keep an eye on this one - and stress test it a bit with large deployments

File: lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py:586]: In general - I wouldn't worry much about the plugin management module, as it requires rework. 

Let's please remember this one. To clarify - I'm referring to the "happy path" when:
- +2 applications bootstrapped and related to the self-signed-certificates 
- both had their certificates issued successfully - with the exception of the app-admin certificate on the "non main orchestrator cluster" 
- with this logic, new units certificates will be reissued _again_

**Review by Mehdi-Bendriss — DISMISSED** [review]
Thanks René - let's please keep an eye in future PRs on the 2 comments highlighted previously. As well as [this issue](https://github.com/canonical/opensearch-operator/issues/417#issuecomment-2313414648)

File: lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py:586]: Yes I understand what you mean, this should be avoided. I will address that in the PR for CA rotation in large deployments.

**Comment by reneradoi** [comment]
> Thanks René - let's please keep an eye in future PRs on the 2 comments highlighted previously. As well as [this issue](https://github.com/canonical/opensearch-operator/issues/417#issuecomment-2313414648)

Yes will do, have linked the PR in the Jira ticket for large deployments.

File: lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py]: has now been removed.

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_tls.py:195]: has been updated

**Comment by juditnovak** [comment]
Unittests are added :-)

I'm waiting for the ongoing issue resolved (the one that @skourta have identified on `certificate-available` event being forever `defer()`-red). Since that's the very same part of the code where the double `app-admin` certificate fetch is happening, it may be a good idea to fix that part of the workflow as well.

**Comment by skourta** [comment]
I met with @reneradoi and explained the issue to him as well as its source. He's on it.

File: lib/charms/opensearch/v0/opensearch_tls.py
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_tls.py]: It's not a direct replacement, as there were two issues:
- if the admin cert was not in the secret yet, this caused `self.tls.update_request_ca_bundle()` to fail
- the check for `self.all_certificates_available()` caused the endless `waiting for TLS to be fully configured` with the `manual-tls operator`

**Review by juditnovak — APPROVED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]

**Comment by juditnovak** [comment]
Important to mention withing the highlights of earlier discussions.
With the new unittests, instead of the original 55% coverage we've ended up at a 62% unittes coverage :-) :-) :-)
image

**Pull request merged**
